### PR TITLE
[WB-5643] Automatically convert logged pandas dataframes to wandb.Tables

### PR DESCRIPTION
**Pull request by dannygoldstein**
https://wandb.atlassian.net/browse/WB-5643

Description
-----------
After attempting to call `wandb.log` on a pandas dataframe, @morganmcg1 reported:

Oh but maybe a curious bug...After getting an error about not being able to log a dataframe, subsequent calls to wandb.log continue to give me the same error, even if what I'm trying to log is valid.
Minimum reproducible example:

```
import wandb
import pandas as pd
wandb.init()
cv_results = pd.DataFrame(data={'test_col': [1,2,3], 'test_col2':[4,5,6]})
wandb.log({'results_df':cv_results})  # Error: We do not support DataFrames in the Summary or History
wandb.log({'results_table': wandb.Table(dataframe=cv_results)})  # Error: We do not support DataFrames in the Summary or History
wandb.log({'tt': 777})  # Error: We do not support DataFrames in the Summary or History
```

Running:
mac 11.3.1 (M1)
python 3.8.8
wandb==0.10.30

This was happening because the `run.history` was not cleared out after this `ValueError` was raised:
image

so the history remained in an interstitial state and attempted to keep flushing the dataframe on future calls to `wandb.log`. This PR attempts to fix this ~~by clearing out the history cache when the ValueError above is raised~~ by automatically converting logged dataframes to `wandb.Table`s.

Testing
-------

Added a unit test.

### CODE REVIEW

File: wandb/sdk/data_types.py [hunk 1]
```diff
@@ -2229,6 +2229,8 @@ def val_to_json(
     typename = util.get_full_typename(val)
 
     if util.is_pandas_data_frame(val):
+        assert run
+        run.history._data = dict()
```
Comment (vanpelt) [inline wandb/sdk/data_types.py:2233]: I think this is problematic.  Users can call `wandb.log` in multiple locations for the same "step" before flushing it.  i.e.

```
wandb.log({"metric_1": N}, commit=False)
wandb.log({"dataframe": df})
```

The above should log both the dataframe and "metric_1".  I believe your code will wipe out "metric_1".  In fact, depending on the keys that are in your log call and the order they happen to be iterated in we drop other keys in the same step.  I'm not sure how to best solve this, even if we only wiped out existing tables in the current histories data we could drop data...

**Review by vanpelt — CHANGES_REQUESTED** [review]

File: wandb/sdk/data_types.py (hunk 1 above)
Comment (dannygoldstein) [reply wandb/sdk/data_types.py:2233]: good point, should we just automatically convert to a wandb table when someone logs a pandas dataframe ?

Comment (vanpelt) [reply wandb/sdk/data_types.py:2233]: I guess for now we can just remove data frames values.  Bonus would be to automatically create wandb tables which seems sane and should fix all cases.

Comment (dannygoldstein) [reply wandb/sdk/data_types.py]: i implemented the second approach but I think it will only work in online mode, since `wandb.Tables` are logged as artifacts. Is this correct and, if so, a problem ?

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -930,7 +930,7 @@ def _history_callback(self, row: Dict[str, Any], step: int) -> None:
         if self._backend:
             not_using_tensorboard = len(wandb.patched["tensorboard"]) == 0
             self._backend.interface.publish_history(
-                row, step, publish_step=not_using_tensorboard
+                row, step, run=self, publish_step=not_using_tensorboard
```
Comment (vanpelt) [inline wandb/sdk/wandb_run.py:933]: Whoa, why was this needed?

**Review by vanpelt — APPROVED** [review]

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (dannygoldstein) [reply wandb/sdk/wandb_run.py]: This was a relic of previous experimentation, removed.
